### PR TITLE
Upgrade to spoon2.0.0-snapshot, android3.0.0 and gradle 4.1

### PR DESCRIPTION
**Pull request by dweebo**
To fix issues discussed in #155 and to prepare for android build tools 3.0.0.

### CODE REVIEW

**Comment by ligi** [comment]
Great initiative - but I have the same problem travis has - as the PR uses a SnapShot of spoon but does not add the SnapShot maven repo. This should help:
```
--- a/build.gradle
+++ b/build.gradle
@@ -33,6 +33,7 @@ repositories {
   mavenLocal()
   google()
   jcenter()
+  maven { url 'https://oss.sonatype.org/content/repositories/snapshots' } 
 }
 

```

**Comment by dweebo** [comment]
Oh nice thanks!  I couldn't figure out where spoon kept snapshots.  I'll update in a bit

**Comment by dweebo** [comment]
3.0.0 is officially out now too so will upgrade to that and test as soon as i can

**Comment by ligi** [comment]
one question (blocks me from using 3.0.0 for now) don't you experience: https://issuetracker.google.com/issues/68275433

**Comment by dweebo** [comment]
@ligi haven't seen that error

**Comment by autonomousapps** [comment]
Hi! Just wondering when this might get merged and published? I also tried to find a snapshot build, but it doesn't seem to be available. https://oss.sonatype.org/content/repositories/snapshots/com/stanfy/spoon/spoon-gradle-plugin/

**Comment by ligi** [comment]
@dweebo thanks for the info and good for you ;-)
@autonomousapps you can get a snapshot via jitpack

**Comment by autonomousapps** [comment]
@ligi can you show me the code for that? Gradle still can't find it. I have:
```
buildscript {
    repositories {
        jcenter()
        maven { url 'https://maven.fabric.io/public' }
        maven { url 'https://plugins.gradle.org/m2/' }
        maven { url 'https://maven.google.com' }
        maven { url 'https://jitpack.io' }
    }
    dependencies {
        // elided
        classpath 'com.stanfy.spoon:spoon-gradle-plugin:2.0.0-SNAPSHOT'
    }
}
```

**Comment by ligi** [comment]
classpath 'com.github.dweebo:spoon-gradle-plugin:5c318830a43c01dc098b745278168e12fb55dedc' should do the trick

**Comment by autonomousapps** [comment]
Thank you, that worked (almost), but now it's complaining it can't find `com.squareup.spoon:spoon-runner:2.0.0-SNAPSHOT`, which it says is required by `com.github.dweebo:spoon-gradle-plugin:5c318830a43c01dc098b745278168e12fb55dedc`

**Comment by ligi** [comment]
you need to add 
 maven { url 'https://oss.sonatype.org/content/repositories/snapshots' } 
to the repositories for your app

**Comment by autonomousapps** [comment]
Yes, sorry! I removed that after you said to use jitpack. All is well now with that re-added to repos block. For anyone else that sees this, here is a complete working example:
```
buildscript {
    repositories {
        maven { url 'https://jitpack.io' }
        maven { url 'https://oss.sonatype.org/content/repositories/snapshots' }
    }
    dependencies {
        classpath 'com.github.dweebo:spoon-gradle-plugin:5c318830a43c01dc098b745278168e12fb55dedc'
    }
}
```

**Comment by jaredsburrows** [comment]
@autonomousapps @ligi When can we get a new release. The last release on this plugin was over a year ago :(

**Comment by ligi** [comment]
@jaredsburrows I cannot tell you as I am not the maintainer ..-)

**Comment by jaredsburrows** [comment]
@ligi Alright.

@autonomousapps Combining it all together:
```
buildscript {
    repositories {
        maven { url "https://jitpack.io" }
        maven { url "https://oss.sonatype.org/content/repositories/snapshots" }
    }

    dependencies {
        classpath "com.squareup.spoon:spoon-runner:2.0.0-SNAPSHOT"
        classpath "com.github.dweebo:spoon-gradle-plugin:5c318830a43c01dc098b745278168e12fb55dedc"
    }
}
```

Both tasks show up again:
```
spoon - Runs all the instrumentation test variations on all the connected devices
spoonDebugAndroidTest - Runs instrumentation tests on all the connected devices for 'debugAndroidTest' variation and generates a report with screenshots
```

**Comment by autonomousapps** [comment]
Cool. I didn't have to add an explicit dependency on the spoon-runner, since the spoon-gradle-plugin pulled that in automatically. :man_shrugging:

**Comment by jaredsburrows** [comment]
@autonomousapps did the output change of the artifacts?

**Comment by autonomousapps** [comment]
@jaredsburrows Good question. I'm currently wrestling with Jenkins to get it to build my project with all the updates. The last build didn't have usual Spoon report, so it's possible the artifact location has changed.

**Comment by autonomousapps** [comment]
@ligi there's something really weird going on with your version of Spoon
1. It's very slow. My test suite has gone from taking ~10 to ~30min.
2. It seems to have misplaced the outputs. I haven't investigated where they are yet, but they're not where they used to me.
3. MOST bizarrely, I have observed that `spoonDebugAndroidTest` is inconsistent with tests run via AS 3.0 AND with tests run with `connectedDebugAndroidTest`. The latter two will instantiate the test class for each test case, whereas Spoon is instantiating the test class just once.

**Comment by ligi** [comment]
@autonomousapps not sure what you mean with my version of spoon ..
I just hinted how to use @dweebo 's branch via jitpack. Personally I am back to build-plugin 2.3.3 and the upstream stanfy plugin as I suffer this bug: https://issuetracker.google.com/issues/68275433 And the workaround does not work for me.

**Comment by autonomousapps** [comment]
Ah, sorry. @dweebo 's version, then :) 

I may have to put 'upgrading' to AGP 3.0 on hold as well. I haven't seen that bug you're facing (I think...), but the ones I have outlined so far are basically show-stoppers.

**Comment by dweebo** [comment]
Yeah that slow down would not be tolerable on my projects either.  I'll try to see what is going on this week, it might be fixed by https://github.com/square/spoon/pull/492

**Comment by jaredsburrows** [comment]
It looks like the task `spoonDebug` changed to `spoonDebugAndroidTest`. It also looks like that the output folder is always in the `build/spoon` folder now.

**Comment by jaredsburrows** [comment]
@dweebo @ligi @autonomousapps Have you all had problems with using this "snapshot" versions?

**Comment by dweebo** [comment]
@jaredsburrows I apologize that I didn't review these changes enough before submitting the PR. 

Today I ran it on a project with around 200 tests.  It went from 26 minutes to 30 minutes so not quite the slowdown you were seeing.  I'm not sure the difference.

I've always seen tasks like `spoonDebugAndroidTest` on my project, I don't recall seeing a `spoonDebug` but I can look into it.  

I was able to set the `baseOutputDir` of the plugin to a custom output directory in my build.gradle and see the spoon output there.

**Comment by jaredsburrows** [comment]
@dweebo I am doing that too. I guess since I used to run `spoonDebug` with `2.3.3` it would out put to a folder called `$baseOutputDir/debug/<files>`. When running `spoonDebugAndroidTest` with `3.0.0` it seems to dump the files into `$baseOutputDir/<files>`.

**Comment by dweebo** [comment]
@jaredsburrows think I fixed the output issue, good catch!

**Comment by autonomousapps** [comment]
The biggest problem I had when testing the snapshot was that it was only instantiating my test classes one time, rather than once per Test method. This caused issues with shared state. I'm certain this is new behavior, because the old spoon plugin ran just fine.

Additionally, my test suite did go from about 15 min to 45. I have about 150 instrumented tests.

**Comment by dweebo** [comment]
I know spoon 2.0 runs instrumentation separately for each test, rather than before it ran instrumentation just once for each connected device and then each test running on that device would have shared-state issues.  But that is the opposite of what you describe so I'm confused.  I honestly haven't looked into how that changes things on my tests yet but I can dig into it.
 
And yeah the speed issue you describe is horrible, that could also be related to the instrumentation changes I talked about above.  How many devices are you running on? Sharding?

**Comment by autonomousapps** [comment]
This is just one device! No sharing.

I can create a little video to demonstrate, if that would help. What I'm
seeing is that each test runs fast, but there's a significant delay BETWEEN
tests. Which is honestly the opposite of what I'd expect if it went from
newing each class many times to newing each class just once.

**Comment by autonomousapps** [comment]
Sorry, on my phone. I of course meant no *sharding.

**Comment by jaredsburrows** [comment]
If this doesn't get in soon, I feel like writing my own spoon gradle plugin. I may start working on the unit/integration tests for it soon.

**Comment by michael-mohemian** [comment]
Pinging the maintainer @roman-mazur on this matter.

**Comment by m2kk** [comment]
I also tried to with above hints and version, but still getting those errors:
`6:24:34 D/ddms: read: timeout
06:24:34 W/PropertyFetcher: TimeoutException getting properties for device 05157df592b63127: null
06:24:34 D/ddms: read: timeout
06:24:34 W/PropertyFetcher: TimeoutException getting properties for device 05157df592b63127: null
06:24:34 D/ddms: read: timeout
06:24:34 W/PropertyFetcher: TimeoutException getting properties for device 05157df592b63127: null
06:24:34 D/ddms: read: timeout
`

and finally it aborts with this one:
`06:24:34 D/ddms: read: timeout
06:24:34 E/Device: Error during Sync: timeout.
2017-11-08 18:24:34 [SDR.run] InstallException while install other apk on device [05157df592b63127]
com.android.ddmlib.InstallException
        at com.android.ddmlib.Device.installPackage(Device.java:918)
        at com.squareup.spoon.SpoonDeviceRunner.run(SpoonDeviceRunner.java:160)
        at com.squareup.spoon.SpoonRunner.runTests(SpoonRunner.java:178)
        at com.squareup.spoon.SpoonRunner.run(SpoonRunner.java:126)
        at com.squareup.spoon.SpoonRunner$run.call(Unknown Source)
        at com.stanfy.spoon.gradle.SpoonRunTask.runSpoon(SpoonRunTask.groovy:184)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:46)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:784)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:751)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:121)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:110)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:60)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:97)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:87)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:626)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:581)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.lang.Thread.run(Thread.java:745)
Caused by: com.android.ddmlib.TimeoutException
        at com.android.ddmlib.AdbHelper.read(AdbHelper.java:879)
        at com.android.ddmlib.AdbHelper.read(AdbHelper.java:848)
        at com.android.ddmlib.AdbHelper.readAdbResponse(AdbHelper.java:220)
        at com.android.ddmlib.AdbHelper.setDevice(AdbHelper.java:972)
        at com.android.ddmlib.SyncService.openSync(SyncService.java:196)
        at com.android.ddmlib.Device.getSyncService(Device.java:520)
        at com.android.ddmlib.Device.syncPackageToDevice(Device.java:949)
        at com.android.ddmlib.Device.installPackage(Device.java:910)
        ... 54 more
2017-11-08 18:24:34 [SR.runTests] [05157df592b63127] Execution done.
`
tried manually installing, works,
tried different devices, no effect, still doesnt work.
tried reverting usb debugging authorization (and resetting), no effect, still doesnt work.

Using Gradle 4.3. Any ideas/help/hints appreciated. Thanks

**Comment by jaredsburrows** [comment]
@michael-mohemian @autonomousapps @ligi @dweebo During KotlinConf, I wrote my own Spoon Gradle Plugin and I included @dweebo's upgrades for AGP 3.0.0 and Spoon 2!

See here: https://github.com/jaredsburrows/gradle-spoon-plugin

It is currently in snapshot only but I actively maintain all of my open source repositories and answer emails very quickly. You should be able to use the same DSL you have now, just a new dependency.

```
buildscript {
  repositories {
    jcenter()
    maven { url "https://oss.sonatype.org/content/repositories/snapshots" } // For Spoon snapshot, until 2.0.0 is released
  }

  dependencies {
    classpath "com.jaredsburrows:gradle-spoon-plugin:1.0.0"
  }
}

apply plugin: "com.android.application"
apply plugin: "com.jaredsburrows.spoon" // or "spoon"
```

Please let me know and try it out. I still need more tests.

**Comment by autonomousapps** [comment]
I'd be happy to give it a try next week! Thanks for the work.

I was at KotlinConf too. Wish I had known - we could have worked on the
plugin together :-)

**Comment by jaredsburrows** [comment]
@autonomousapps That's great to hear that you went as well! We still can work on it together, I will actively accept PRs as well as do releases more often.

**Comment by michael-mohemian** [comment]
@jaredsburrows wow that's great, thank's for your effort!

**Comment by jaredsburrows** [comment]
@michael-mohemian No problem. For some reason, you might have to use the SNAPSHOT until jcenter syncs my releases.

**Comment by Kusand** [comment]
@jaredsburrows Thanks - I'll switch over to your version in the near future and get you some feedback!

**Comment by ligi** [comment]
https://github.com/square/spoon/commit/53a63fe4a5182773a05c21c8c494cf754c793bf3 was just merged - @jaredsburrows can you ping when a version of your plugin contains this?

**Comment by jaredsburrows** [comment]
@ligi Support added here in `1.1.11-SNAPSHOT`.

**Comment by jaredsburrows** [comment]
@dweebo @ligi @autonomousapps Are you all seeing this? https://github.com/square/spoon/issues/498

No xml junit reports are being generated.

**Comment by jaredsburrows** [comment]
@ligi Can you verify that "single instrumentation" is working for you?

Also, @dweebo @autonomousapps are you all seeing what I am seeing?

**Comment by ligi** [comment]
@jaredsburrows sorry - did not yet find time to try - also a bit reluctant using such snapshot versions as I want my builds to be reproducible as possible - as far as I see this way the plugin could change from build to build - which means the build could pass one time and fail another time.
But might try to use it via jitpack with a referenced commit-hash soon - should also work. PS does your plugin support the spoon flag --fail-if-no-device-connected ?

**Comment by jaredsburrows** [comment]
@ligi I need to publish a new release but the SNAPSHOT is stable and does support `fail-if-no-device-connected`. I also have updated it to support "single instrumentation". Regardless of snapshot or not, you still have to use the SPOON 2.0.0 SNAPSHOT for this plugin to work.

**Comment by ligi** [comment]
Ah right - dam - do you have any idea when spoon 2.0.0 will be released?

**Comment by jaredsburrows** [comment]
@ligi No :(

**Comment by hram** [comment]
Hi all
Can you check SNAPSHOT how it works if specify package for running tests only from this package?
In my project when I set instrumentation arguments like this

> instrumentationArgs = ["package=my.app.package.name"]

I haw exception in logcat

> E/AndroidJUnitRunner: Fatal exception when running tests
                      java.lang.IllegalArgumentException: Ambiguous arguments: cannot provide both test package and test class(es) to run
                          at android.support.test.internal.runner.TestRequestBuilder.validate(TestRequestBuilder.java:828)
                          at android.support.test.internal.runner.TestRequestBuilder.build(TestRequestBuilder.java:789)
                          at android.support.test.runner.AndroidJUnitRunner.buildRequest(AndroidJUnitRunner.java:487)
                          at android.support.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:373)
                          at my.package.app.runner.UnlockDeviceAndroidJUnitRunner.onStart(UnlockDeviceAndroidJUnitRunner.java:44)
                          at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2075)

**Comment by jaredsburrows** [comment]
@hram Yes. This is because by default Spoon 2 runs in separate instrumentation. So you need to use the `singleInstrumentationCall = true` from https://github.com/jaredsburrows/gradle-spoon-plugin to make it work. Or you can use annotations such as `@SmallTest`.

**Comment by hram** [comment]
@jaredsburrows thank you very much!
I will try
As workaround I configure Suite with all testes what I needed
And configure spoon with parameter 'className' to run this Suite

**Comment by jaredsburrows** [comment]
@hram Great! It really should be reported to spoon. As it is a limitation to `adb shell am instrumentation`.

File: build.gradle [hunk 1]
```diff
@@ -31,15 +31,20 @@ targetCompatibility = 1.6
...
 
 dependencies {
   compile gradleApi()
   compile localGroovy()
 
-  compile 'com.squareup.spoon:spoon-runner:1.7.1'
+  compile 'com.squareup.spoon:spoon-runner:2.0.0-SNAPSHOT'
 
-  compile 'com.android.tools.build:gradle:1.5.0'
+  compile('com.android.tools.build:gradle:3.0.0') {
```
Comment (jimxj) [inline build.gradle]: Could you update it to 3.0.1 ?

Comment (jaredsburrows) [reply build.gradle]: Try using https://github.com/jaredsburrows/gradle-spoon-plugin.

Comment (jimxj) [reply build.gradle]: Actually I am trying :)

**Comment by dweebo** [comment]
@jimxj I upgraded to 3.0.1 and fixed an issue with a recent change to spoon snapshot 2.0.0

**Comment by jaredsburrows** [comment]
@dweebo I love the map.

**Comment by hram** [comment]
@jaredsburrows can you help me?
I was trying to set package name as

> instrumentationArgs = ["package my.app.testspoon"]

and set parameter as you sad

> singleInstrumentationCall = true

but spoon task execuded fith error

> java.lang.NoSuchMethodError: com.squareup.spoon.SpoonRunner$Builder.setInstrumentationArgs(Ljava/util/List;)Lcom/squareup/spoon/SpoonRunner$Builder;
        at com.jaredsburrows.spoon.SpoonTask.spoonTask(SpoonTask.kt:73)

I changed params as 

> instrumentationArgs = ["package=my.app.testspoon"] 

and no luck (

**Comment by jaredsburrows** [comment]
@hram Sorry. I need to update the documentation. Your `instrumentationArgs = ["package my.app.testspoon"]` should be `instrumentationArgs = ["package=my.app.testspoon"]`.

I need to support both `List` and `Map`.

**Comment by hram** [comment]
@jaredsburrows but as I sad

> I changed params as
> instrumentationArgs = ["package=my.app.testspoon"]
> and no luck (

**Comment by jaredsburrows** [comment]
@hram Maybe it is a bug. Can you post an issue in that projects issues?

**Comment by hram** [comment]
@jaredsburrows yes I can ) 
https://github.com/jaredsburrows/gradle-spoon-plugin/issues/8

**Comment by Igorshp** [comment]
Is anythig blocking this from being merged?

**Pull request closed without merging**